synapse/storage/data_stores/main/schema/delta/56/event_labels.sql

```diff
@@ -0,0 +1,22 @@
+/* Copyright 2019 The Matrix.org Foundation C.I.C.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+CREATE TABLE IF NOT EXISTS event_labels (
+    event_id TEXT,
+    label TEXT,
+    PRIMARY KEY(event_id, label)
+);
+
+CREATE INDEX event_labels_label_idx ON event_labels(label);
```

Comment:
i'm not sure this index is useful?

Reply:
Given we filter on labels in the `WHERE` clause when paginating, cf

https://github.com/matrix-org/synapse/blob/fe51d6cacf6e1a2da5fc3589d0bc4118342b33dd/synapse/storage/data_stores/main/stream.py#L232-L238

I'd have thought indexing on the labels would be useful.

Reply:
This has been fixed by adding the room ID to the table. On top of that, on @erikjohnston's advice, I've also added the topological order of events to the table, so that the engine doesn't have to load the entire table to work (given the SQL query orders by topological ordering and limits).

The resulting indexed tuple is `(room_id, label, topological_ordering)`.